### PR TITLE
fix: Improving validate_task_to_prompt_name logs and error messages

### PR DESCRIPTION
**Pull request by RyanMullins**
## Problem

As released, the `Wrapper.validate_task_to_prompt_name()` function raises a `KeyError` if it encounters any key in the prompts dictionary that doesn't match a specific formatting requirement, even if the provided prompts dictionary contained _some_ valid keys. 

In most cases, `Wrapper` subclasses would pass this `KeyError` through and cause the MTEB wrapper initialization or evaluation run to fail.

The `SentenceTransformerWrapper` behaved differently in that it would catch the error raised by `Wrapper.validate_task_to_prompt_name()` and set the model prompts to `None`, allowing the evaluation to run but without any of the prompts that would make the model perform as expected. 

The `ColBERTWrapper` subclass was trying to catch a `ValueError` that did not seem like it would ever come up.

The problems documented above were observed when testing a `SentenceTransformer` model that included some custom task names in its `prompts` configuration. Sentence Transformers does not place restrictions on the values of keys in the `prompts` dictionary, so: 

* Having task names that are unknown to MTEB should be acceptable for the model given the Sentence Transformers best practices; and
* Ideally MTEB would use the prompts it does recognize instead of refusing the use any prompts if it encounters one unrecognized prompt. 

## Proposed Solution

I propose refactoring `Wrapper.validate_task_to_prompt_name()` to support options for:

1.  Raising a `KeyError`, for backward compatibility; or 
1.  Filtering to the valid prompts for MTEB and reporting the invalid prompts without raising.

In either case of the above, `Wrapper.validate_task_to_prompt_name()` should record the entire set of problems with the keys being validated, so that users can see and address all issues at once instead of having to iteratively run the configured model prompts through `Wrapper.validate_task_to_prompt_name()` to fix them all.

This PR: 

* Refactors `Wrapper.validate_task_to_prompt_name()` to address the above
* Refactors most `Wrapper` subclasses to 1) move a `None`-ness check inside the initializer function to the `Wrapper.validate_task_to_prompt_name()` function and 2) use the "raises" variant of the refactored `Wrapper.validate_task_to_prompt_name()` function.
* Refactors the `SentenceTransformerWrapper` class to use the "filters and reports" variant of the refactored `Wrapper.validate_task_to_prompt_name()` function.
* Updates existing unit tests as appropriate
* Adds unit tests for the "filters and reports" variant of the `Wrapper.validate_task_to_prompt_name()` function.

### CODE REVIEW

File: mteb/models/sentence_transformer_wrapper.py [hunk 1]
```diff
@@ -39,28 +39,35 @@ def __init__(
...
+
+        self.model_prompts, invalid_prompts = self.validate_task_to_prompt_name(
+            model_prompts, raise_for_invalid_keys=False
+        )
+
+        if invalid_prompts:
+            logger.warning(
+                "Some prompts are not in the expected format and will be ignored. Problems:\n\n"
+                f"{'\n'.join(invalid_prompts)}"
+            )
```
Comment (Samoed) [inline mteb/models/sentence_transformer_wrapper.py]: I think this can be still in one line. I think it's a bit unexpected, when logs are multiline

File: mteb/models/wrapper.py [hunk 2]
```diff
@@ -65,29 +66,95 @@ def get_prompt_name(
...
 
     @staticmethod
+    @overload
     def validate_task_to_prompt_name(
-        task_to_prompt_name: dict[str, str] | None,
-    ) -> dict[str, str] | None:
-        if task_to_prompt_name is None:
-            return task_to_prompt_name
+        task_to_prompt: dict[str, str] | None,
+        raise_for_invalid_keys: Literal[True] = True,
```
Comment (Samoed) [inline mteb/models/wrapper.py:72]: Nice hack, but I'm not sure about it, because it's a bit tricky

Comment (RyanMullins) [reply mteb/models/wrapper.py:72]: Yeah... makes me really wish Python had a better typing system. Open to suggestions.

File: mteb/models/sentence_transformer_wrapper.py (hunk 1 above)
Comment (RyanMullins) [reply mteb/models/sentence_transformer_wrapper.py]: Fixed.

File: tests/test_reproducible_workflow.py [hunk 3]
```diff
@@ -67,11 +74,43 @@ def test_validate_task_to_prompt_name(task_name: str | mteb.AbsTask):
...
+    ],
+)
+def test_validate_task_to_prompt_name_fails_and_raises(
+    task_prompt_dict: dict[str, str],
+):
     with pytest.raises(KeyError):
-        Wrapper.validate_task_to_prompt_name(
-            {"task_name": "prompt_name", "task_name-query": "prompt_name"}
-        )
+        Wrapper.validate_task_to_prompt_name(task_prompt_dict)
```
Comment (Samoed) [inline tests/test_reproducible_workflow.py:97]: Can you add test for `None` input prompts?

**Review by Samoed — COMMENTED** [review]
Overall I think this is good changes. Thanks!

File: tests/test_reproducible_workflow.py (hunk 3 above)
Comment (RyanMullins) [reply tests/test_reproducible_workflow.py:97]: Done

File: mteb/leaderboard/benchmark_selector.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 from dataclasses import dataclass
 
 import gradio as gr
+from build.lib.mteb.benchmarks.benchmarks import MTEB_multilingual
```
Comment (Samoed) [inline mteb/leaderboard/benchmark_selector.py]:
```suggestion
```

**Review by KennethEnevoldsen — APPROVED** [review]
I agree with @Samoed here I think this is a good change.
